### PR TITLE
[12.0][IMP] account_multicurrency_revaluation

### PR DESCRIPTION
**Pull request by alexey-pelykh**
Things done:
 * [FIX] #126, #117: bank account revaluations are not grouped by partner
 * [FIX] Bank account revaluations are not performed if currency of account matches company currency
 * [IMP] When revaluation is performed in the past, force-execute reversal (if needed)
 * [IMP] Exchange rate is rounded during representation in same way as Odoo does
 * [IMP] Float comparison operations take currency rounding into account
 * [FIX] README cleanup

### CODE REVIEW

**Comment by alexey-pelykh** [comment]
@gdgellatly @pedrobaeza @yvaucher @hbto @fclementic2c @gurneyalex @matthieuchoplin it would be awesome if you could review this PR.

**Comment by hbto** [comment]
@luisg123v Could provide one of your insightful Technical Reviews on this MR?

Regards.

**Comment by alexey-pelykh** [comment]
Any feedback?

**Comment by hbto** [comment]
@alexey-pelykh Coverage has decreased for the overall projec. Though for a slim margin.

File: account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py [hunk 1]
```diff
@@ -273,83 +254,102 @@ def revaluate_currency(self):
...
-                        sums
+                        lines
                     )
                     created_ids.extend(new_ids)
 
+        # In case revaluation date is before today, it's safe to run reversing
+        # w/o waiting tomorrow, since otherwise it would cause confusion when
+        # revaluating historical entries for multiple years within one day.
+        if self.journal_id.company_id.reversable_revaluations \
+                and self.revaluation_date < fields.Date.today():
```
Comment (hbto) [inline account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py]: use `fields.Date.context_today(self)`

**Comment by hbto** [comment]
> [FIX] #126, #117: bank account revaluations are not grouped by partner

Bank revaluation can be easily done by just taking the full balance of the account at one point, the full amount_currency at that point, multiplying the amount_currency times the rate of that currency against local/company currency and making the subtraction.
Having more lines per partner to revaluate the banks in my opinion is just a cosmetic change, it results to the accounting people too much hassle to analyze when only a line with Revaluation Gain/Loss per bank is needed.

Maybe if that is optional and not mandatory for me it could a more desirable change.

> [FIX] Bank account revaluations are not performed if currency of account matches company currency
+1 

> [IMP] When revaluation is performed in the past, force-execute reversal (if needed)
+1

> [IMP] Exchange rate is rounded during representation in same way as Odoo does
+1

> [IMP] Float comparison operations take currency rounding into account
+1

> [FIX] README cleanup
+1

**Comment by alexey-pelykh** [comment]
@hbto 

> Having more lines per partner to revaluate the banks in my opinion is just a cosmetic change, it results to the accounting people too much hassle to analyze when only a line with Revaluation Gain/Loss per bank is needed.

Bank accounts are not grouped by partner, so that's not the case.

File: account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py (hunk 1 above)
Comment (alexey-pelykh) [reply account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py]: Valid point, amended

**Comment by hbto** [comment]
> @hbto
> 
> > Having more lines per partner to revaluate the banks in my opinion is just a cosmetic change, it results to the accounting people too much hassle to analyze when only a line with Revaluation Gain/Loss per bank is needed.
> 
> Bank accounts are not grouped by partner, so that's not the case.

Excellent.

Let's merge it.

Regards 

:+1:

**Comment by hbto** [comment]
@pedrobaeza Could you please check on this MR?

Regards.

**Comment by hbto** [comment]
@alexey-pelykh 


However please do take care of the Coverage thus allowing this PR to merge smoothly.

Regards.

**Comment by pedrobaeza** [comment]
Travis is red

**Comment by alexey-pelykh** [comment]
@pedrobaeza on it, context_today() requires change in tests as well

**Comment by alexey-pelykh** [comment]
@pedrobaeza fixed

**Comment by pedrobaeza** [comment]
OK, but I'm not able to do a functional review.

**Comment by alexey-pelykh** [comment]
at least code review would be great

**Comment by hbto** [comment]
@pedrobaeza & @alexey-pelykh 

I have test it Functionally and It Looks Good to me.

Using this Journal Entries:
Screen Shot 2021-01-16 at 23 26 39

The Result is this One

For Journal Items with Currency Set the result is:
Screen Shot 2021-01-16 at 23 26 10

For the whole set of Item including any currency:
Screen Shot 2021-01-16 at 23 27 27

Regards.

**Comment by hbto** [comment]
@alexey-pelykh the only missing thing here is the slim coverage that needs to be overcome.

Regards.

**Comment by hbto** [comment]
@alexey-pelykh There is another request could you put in two commits the changes.

One for the `account_multicurrency_revaluation` module you are fixing.
Another for the`account_fiscal_year_closing` module where a test is being fixed.

Regards.

**Comment by alexey-pelykh** [comment]
@hbto moved fiscal tests fixing into #151

**Comment by alexey-pelykh** [comment]
@hbto codecov is now happy too

**Comment by hbto** [comment]
@alexey-pelykh Cheers,

CI & CD can sometimes be demanding.

Cheers 

@pedrobaeza (or someone else) Could you please take care this MR?

Regards.

**Review by hbto — APPROVED** [review]
👍  Looks Good To Me.

Regards.

**Comment by alexey-pelykh** [comment]
@sergiocorato would it be possible to request your review for this one?

**Comment by alexey-pelykh** [comment]
Can we do something about this PR finally? :)

**Comment by alexey-pelykh** [comment]
@OCA/accounting-maintainers may I request some attention to this?

**Comment by alexey-pelykh** [comment]
@kittiu @sbidoul any luck getting a review? :)

**Review by gaikaz — APPROVED** [review]
Nice work @alexey-pelykh !
Patched code LGTM

Functional review of the fixes was done by my colleague.
This does fix an issue our client has.

Any chance of this getting forwarded to v13 when it gets merged?

**Comment by alexey-pelykh** [comment]
@OCA/accounting-maintainers can we get this merged?

@gaikaz glad to hear that! Regarding v13: not on the schedule right now, but feel free to reach me out via email to discuss the priority.

**Pull request merged**